examples/nucleo_f439zi/can_m2515/main.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2013, Kevin Läufer
+ * Copyright (c) 2013-2017, Niklas Hauser
+ * Copyright (c) 2016, Raphael Lehmann
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#include <modm/board.hpp>
+#include <modm/platform/spi/spi_master_1.hpp>
+#include <modm/platform/spi/spi_master_2.hpp>
+#include <modm/processing/protothread.hpp>
+#include <modm/driver/can/mcp2515.hpp>
+
+#define SENDER 1
+
+// Set the log level
+#undef MODM_LOG_LEVEL
+#define MODM_LOG_LEVEL modm::log::DEBUG
+
+// If you use a different SPI instance, you may have to also choose different
+// GPIOs to connect to.
+using Cs = GpioOutputA4;
+using Mosi = GpioOutputB5;
+using Miso = GpioInputB4;
+using Sck = GpioOutputB3;
+using Int = GpioInputC7;
+using SpiMaster = SpiMaster1;
+// Note that you can also use a bit-banged SPI driver as a drop-in replacement
+// using SpiMaster = BitBangSpiMaster<Sck, Mosi, Miso>;
+
+// Default filters to receive any extended CAN frame
+FLASH_STORAGE(uint8_t canFilter[]) = {
+	MCP2515_FILTER_EXTENDED(0),  // Filter 0
+	MCP2515_FILTER_EXTENDED(0),  // Filter 1
+
+	MCP2515_FILTER_EXTENDED(0),  // Filter 2
+	MCP2515_FILTER_EXTENDED(0),  // Filter 3
+	MCP2515_FILTER_EXTENDED(0),  // Filter 4
+	MCP2515_FILTER_EXTENDED(0),  // Filter 5
+
+	MCP2515_FILTER_EXTENDED(0),  // Mask 0
+	MCP2515_FILTER_EXTENDED(0),  // Mask 1
+};
+
+modm::Mcp2515<SpiMaster, Cs, Int> mcp2515;
+
+class MyTask : modm::pt::Protothread
+{
+public:
+	MyTask() :
+		message_{0x1337},
+		i_{0},
+		j_{0}
+		{}
+
+	bool
+	run()
+	{
+		PT_BEGIN();
+
+		MODM_LOG_INFO << "Initializing mcp2515 ..." << modm::endl;
+		// Configure MCP2515 and set the filters
+		initialized_ = mcp2515.initialize<8_MHz, 500_kbps>();
+		MODM_LOG_INFO << "Success: " << initialized_ << modm::endl;
+		mcp2515.setFilter(modm::accessor::asFlash(canFilter));
+		MODM_LOG_INFO << "Running ... " << modm::endl;
+#if SENDER == 0
+		while (initialized_)
```

Comment:
Weird control flow.

I'd prefer a more conventional control flow, e.g.:

```c++
while (! mcp2515.initialize<8_MHz, 500_kbps>()) {
  PT_YIELD();
  // maybe some delay?
}
// ...
while (1) {
  // ...
}
```

Reply:
i changed it to be like this ... for whatever reason :)